### DISCUSSION TITLE
Roadmap?, Squirrel.Windows?, Clowd.Squirrel?

### DISCUSSION THREAD

**Discussion by dopry** [Squirrel/Squirrel.Windows#1811] [General]
@robmen @caesay, @anaisbetts  first I just wanted to say I appreciate all the work you all appear to be doing to get Squirrel.Windows back in motion.  You are the current maintainers? 

I was wondering if there is any sort of roadmap or priorities you all have in mind for Squirrel.Windows. From the commit history it seems like there is an effort underway to get CI/CD running on github actions. 

@caesay appears to be doing some excellent work on Clowd.Squirrel, there are quite a few features there that are very attractive. Are there plans to attempt to backport any of those? 

Are there any particular types of support you need from the wider community of users and developers? 

I know my personal itch is mostly related to troubles I face with code signing.  I can personally commit time to that effort in the form of code, code review, testing, or documentation as needed.

**Comment by caesay** [comment]
I originally offered to contribute my changes back incrementally to this project, however when that was declined I opted to diverge more significantly.

v2 of my fork shares some aspects with this repo. As of v3, the cross-platform rework, there is _very little_ shared code left in my fork, and back-porting any changes would be unmergable and manual.

Off the top of my head, some of the improvements of my fork:
- Streamlined Setup.exe architecture, faster, smaller, can be built on other OS's without windows or wine.
- Properly support dotnet core, single file apps etc. Squirrel.Windows does not do this well.
- Splash window supports any kind of images and appears faster. Also shows install progress (must-have for slow computers imo!). Also multi-mon dpi supported nicely.
- Supports installing netframework, netcore, or vcredist while installing. Also can update frameworks while updating.
- Apps are installed to a `current` dir, and don't move with updates. (v3 only). 
- Delta updates are way faster to build and apply (v3 mostly).
- LZMA compression = smaller packages (v3 only)
- Self-contained and zero dependencies on external frameworks at runtime
- A cross platform global dotnet tool for running Squirrel `csq` that checks the PackageVersion in your project and bundles your app using the correct version of the command line tools (eg. If you reference Clowd.Squirrel 2.9.40, it will use Squirrel.exe 2.9.40 to package. If you reference Clowd.Squirrel 3.0.* it will use the appropriate csq.dll to package as there is no exe). It also supports cross-compiling (eg release your windows program on mac or linux).
- Proper package uploading/syncing commands make it a breeze to deploy your releases and apply retention policies
- Minimal other package dependencies (only SharpCompress)
- Multi-targeted nuget package (netstandard, net461, net5, net6) with proper `[SupportedOS]` annotations
- macOS support (v3 only)
- Way better logging; command line tools (eg. Squirrel.exe, csq) log only to console, have colored messages. Installed tools (eg. Update.exe, UpdateMac) apply log retention policies so log files stay at 1mb instead of growing to >1gb as they do with Squirrel.Windows
- Temp files are cleaned up properly next time Squirrel runs rather than them being left around forever even if they are left behind by a crash. And stored properly in the Temp directory on Windows and Mac so they will get cleaned up by the OS if disk space is needed or according to user/cron settings.
- Probably lots more smaller fixes I can't remember.

All of that doesn't necessarily capture how much easier `Clowd.Squirrel` is to use, but that has been a huge focus for me. It should just work out of the box, be fast, have commands for all the common use-cases and be easy to automate. Messages should be helpful and useful to even a beginner. Example: forget to add `--self-contained` when building, and forget to add `-f net6.0`? you'll get a warning when building releases that suggests you add `-f net6.0` (or whatever framework you're actually missing).

Future roadmap-ish:
- Add dll-based extensibility in Update.exe, allow custom splash window UI's, EULA's etc.
- Deprecate releasify command at some point (there is a 'pack' command which takes a build output)
- Add macOS installer packages (.pkg)
- Add support for bootstrapping the `WindowsAppSDK` when that gets a little more stable.

I provide support on Discord and am happy to provide help to anyone who is looking to use my fork, or I can also support on Discord anyone who is looking to manually back-port changes to this project and would like a steer from me.

**Comment by robmen** [comment]
@dopry yes, I am slowly assuming the role of active maintainer.

I ran all the ideas I had about how to take Squirrel forward by @anaisbetts and incorporated all of her guidance into my thinking. I'm now working through a few other big decisions at work (FireGiant) before laying down a concrete plan for Squirrel v3. I don't expect there will be any surprises in the first draft for v3 since I've mostly been looking at the many forks that sprouted up over the last year or two to resolve their biggest pain points.

As you noticed, while I finalize those big decisions at work, I am investing in infrastructure improvements that benefit Squirrel v2 and (I expect) v3. My next "big" coding task there is to implement the integration tests fixture. I believe this is very important and it should be a fantastic place for other developers to contribute so future Squirrel development does not regress their key scenarios.

Now I know I did not provide a lot of concrete answers about Squirrel v3 in the above. Clarity is coming as soon as I resolve these few big decisions at work (can you tell what is top of mind for me right now? 😄).
